ТЗ 13 ShareIt часть 1

RomanAntipenko opened the pull request:
feat: создай классы Item и User
feat: создай ItemDTO и UseDto для работы в контролерах и передачи их фронту вместо Item и User
feat: Создай интерфейсы: сервисов, контроллеров и репо для них (и аналогично классы) соответственно
feat: Создай обработчик ошибок и ErrorResponse для конкретезирования эксепшенов
feat: Создай интерфесы для группировки @Validated
feat: Добавь мапперы для моделей и дто
feat: Напиши тесты

RomanAntipenko committed:
feat: создай классы Item и User
feat: создай ItemDTO и UseDto для работы в контролерах и передачи их фронту вместо Item и User
feat: Создай интерфейсы: сервисов, контроллеров и репо для них (и аналогично классы) соответственно
feat: Создай обработчик ошибок и ErrorResponse для конкретезирования эксепшенов
feat: Создай интерфесы для группировки @Validated
feat: Добавь мапперы для моделей и дто
feat: Напиши тесты

RomanAntipenko committed:
style: Подправь стилистические ошибки

stfaustoff commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+        return ItemMapper.mapToDto(itemService.patchItem(item));
+    }
+
+    /*@GetMapping
```
Закомментированный код перед PR лучше удалять.

stfaustoff commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.stream.Collectors;
+
+/**
+ * TODO Sprint add-controllers.
```
Вроде как TODO выполнено, можно и убрать.

stfaustoff commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,35 @@
+import org.springframework.stereotype.Component;
+import ru.practicum.shareit.item.model.Item;
+
+@Component
```
Данный класс не надо делать компонентом, у него все методы статические, соответственно внедряться как зависимость ему никуда ненадо.

stfaustoff commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+        return itemRepository.getItemByItemId(itemId);
+    }
+
+    public boolean validateUserId(long userId) {
```
По идее этот метод внутренний, раз он отсутствует в интерфейсе и его не вызывают снаружи.
Тогда лучше сделать его private

stfaustoff commented on src/main/java/ru/practicum/shareit/item/validations/FirstlyItemValidation.java:
```diff
@@ -0,0 +1,4 @@
+package ru.practicum.shareit.item.validations;
+
+public interface FirstlyItemValidation {
```
Не совсем понял, зачем нужен этот интерфейс, а так же SecondaryItemValidation.

RomanAntipenko replied:
Для группировки валидации потому что иногда надо валидировать по одному а иногда по другому или как то можно по другому?

stfaustoff replied:
Я просто не нашел мест его использования.

stfaustoff commented on src/main/java/ru/practicum/shareit/user/repository/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,73 @@
+    private final Map<Long, User> userMap;
+
+    @Setter
+    private long idGenerator;
```
_Предложение_
Ради разнообразия можно посмотреть в стороне AtomicLong

stfaustoff commented on src/main/java/ru/practicum/shareit/user/service/UserService.java:
```diff
@@ -0,0 +1,17 @@
+
+    Collection<User> getAllUsers();
+
+    User patchUser(User user);
```
Более используемый оборот - update, как на уровне команд sql, ведь все таки мы обновляем объект, а не исправляем его.
Именование имеет важную роль, так как является первой частью документации с которой сталкивается другой разработчик.

stfaustoff commented on src/main/java/ru/practicum/shareit/user/validations/FirstlyUserValidation.java:
```diff
@@ -0,0 +1,4 @@
+package ru.practicum.shareit.user.validations;
+
+public interface FirstlyUserValidation {
```
К этому интерфейсу тоже есть вопрос.

RomanAntipenko committed:
style: Убери закомментированный код
fix: Используй генерацию новых id через АтомикЛонг
style: Измени патч на апдейт в названиях методов везде кроме контроллеров
fix: Убери аннотацию component из мапперов
fix: Подправь тесты

RomanAntipenko committed:
style: как же я упускаю эти анюзд импортс

RomanAntipenko commented:
Исправил все кроме интерфейсов, они для группировки валидации потому что иногда надо валидировать по-одному, а иногда по-другому
